src/model.rs

```diff
@@ -724,6 +724,124 @@ impl Collections {
+    /// means authorized).
+    ///
+    /// Example 1:
+    /// ----------
+    ///       out          in   out         in
+    ///        X    SP1    |    ▲    SP2    X
+    ///        X           ▼    |           X
+    ///  VJ:1   08:00-09:00      10:00-11:00
+    ///  VJ:2                    10:00-11:00      14:00-15:00
+    ///                         X           ▲    |           X
+    ///                         X           |    ▼   SP3     X
+    ///                         out         in   out         in
+    ///                         |- Stay-In -|
+    ///
+    /// Example 2:
+    /// ----------
+    ///                                     (1)  (2)
+    ///       out          in   out         in   out         in   out         in
+    ///        X    SP1    |    ▲    SP2    X    X    SP3    |    ▲   SP4     X
+    ///        X           ▼    |           X    X           |    |           X
+    ///  VJ:1   08:00-09:00      10:00-12:00     X           ▼    |           X
+    ///  VJ:2                                     11:00-13:00      13:00-14:00
+    ///                         |---------- Stay In ---------|
+    ///
+    /// Note the overlap between the departure time of the last stop point SP2
+    /// of VJ:1 and the arrival time of the first stop point SP3 of VJ:2. In
+    /// this case, we still apply the default rule.
+    ///
+    ///
+    /// Example 3:
+    /// ----------
+    ///                                     (1)  (2)
+    ///       out          in   out         in   out         in   out         in
+    ///        X    SP1    |    ▲    SP2    |    ▲    SP3    |    ▲   SP4     X
+    ///        X           ▼    |           ▼    |           |    |           X
+    ///  VJ:1   08:00-09:00      10:00-11:00     |           ▼    |           X
+    ///  VJ:2                                     12:00-13:00      14:00-15:00
+    ///                         |---------- Stay In ---------|
+    ///
+    /// Example 3 is the only case were we allow specific pick-up and drop-off
+    pub fn enhance_pickup_dropoff(&mut self) {
+        let mut vj_idxs: Vec<Idx<VehicleJourney>> = self
+            .vehicle_journeys
+            .iter()
+            // Filtering out vehicle journeys without any stop time allowing us
+            // below to call stop_times[0] and stop_times[stop_times.len() - 1]
+            // safely.
+            .filter(|(_, vj)| !vj.stop_times.is_empty())
+            .map(|(idx, _)| idx)
+            .collect();
+        // Sort the vehicule journeys by block_id, and if equal, order them by
+        // departure time.
+        vj_idxs.sort_by(|vj1_idx, vj2_idx| {
+            let vj1 = &self.vehicle_journeys[*vj1_idx];
+            let vj2 = &self.vehicle_journeys[*vj2_idx];
+            match vj1.block_id.cmp(&vj2.block_id) {
+                Ordering::Equal => vj1.stop_times[0]
+                    .departure_time
+                    .cmp(&vj2.stop_times[0].departure_time),
```

Comment:
I think sorting is not good. It must be done on the minimal positive duration between the last schedule of the first vj ant the first schedule of the second vj.

Reply:
If they're part of the same `block_id` (condition `Ordering::Equal` to end up on this line of the code), then ordering by starting time should be enough I believe. Remember that in any case, I'm checking for overlapping times later in this function.

If I understand correctly what you propose, I believe it will actually cause problems. Imagine a VJ_A that starts at 10:00 et arrives at 14:00 then a VJ_B that starts at 11:00 et arrives at 13:00. Following your rule, the sorting will be different depending of if VJ_A is `vj1` or is VJ_A is `vj2` (or said differently, depending on the order they appears in `trips.txt`).

In first case, `vj1.stop_times[vj1.stop_times.len() - 1] = 14:00` (last schedule of first vj) is bigger than `vj2.stop_times[0] = 11:00` (first schedule of second vj), therefore final order will be `VJ_B -> VJ_A`. 

In second case, `vj1.stop_times[vj1.stop_times.len() - 1] = 13:00` (last schedule of first vj) is bigger than `vj2.stop_times[0] = 10:00` (first schedule of second vj), therefore final order will be `VJ_A -> VJ_B`.

Reply:
After a discussion with @patochectp, the following case would cause a problem.

```    
          SP0           SP1           SP2           SP3            
           |             |             |             |             
VJ1   09:00-10:00   10:00-11:00        |             |             
           |             |             |             |             
VJ2        |             |        12:00-13:00   14:00-15:00        
           |             |             |             |             
VJ3   09:30-10:30   10:30-11:30        |             |             
           |             |             |             |             
VJ4        |             |        12:30-13:30   14:30-15:30        

```
In this case, VJ1 and VJ2 should be considered part of a prolongation trip pair; VJ3 and VJ4 should be considered part of another prolongation trip pair. But in terms of modelization, they might have the same `block_id`.

The rest of the implemented algorithm relies on the fact that successive vehicle journeys in a prolongation trip pair, should be successive after the sort. Therefore, a correct sort is VJ1 < VJ2 < VJ3 < VJ4. But the sort function I implemented so far will use `stop_times[0].departure_time` and the resulting sort will be VJ1 < VJ3 < VJ2 < VJ4 which is incorrect.

Reply:
We decided to not support stay-in situations for vehicle journeys with different validity patterns. The above case is most likely to make sense with different validity patterns. If the 4 vehicle journeys above have the same validity pattern, then VJ1/VJ2 will most likely have a different `block_id` than VJ3/VJ4.

To answer to the original [comment](https://github.com/CanalTP/transit_model/pull/703#discussion_r486838087) about the sort, using `departure_time` as a sort-key is not perfect but should be enough if we ignore validity patterns. The logical implication here might not be obvious (it's a shortcut :stuck_out_tongue:), but source code has been documented for more details about that reasoning.

I'm resolving this conversation.